File: 백준/Silver/9184. 신나는 함수 실행/신나는 함수 실행.java

import java.io.*;
import java.util.*;

public class Main {

    // static BufferedWriter bw = new BufferedWriter(new
    // OutputStreamWriter(System.out));
    static int a,b,c;
    static int[][][] dp = new int[21][21][21];
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        // BufferedReader br = new BufferedReader(new FileReader("input.txt"));
        //StringTokenizer st = new StringTokenizer(br.readLine());
        StringTokenizer st;
        while (true) {
            st =new StringTokenizer(br.readLine());
            a = Integer.parseInt(st.nextToken()); 
            b = Integer.parseInt(st.nextToken()); 
            c = Integer.parseInt(st.nextToken());
            if(a==-1 && b==-1 && c==-1) break;

            System.out.printf("w(%d, %d, %d) = %d\n",a,b,c,w(a,b,c));
        }
    }
    /** 메모이제이션 
     *  dp[][][] 배열에 각각 a,b,c에 관한 값들을 집어넣는다(memo)
     * **/
    static int w(int a, int b, int c) { 
        if(a>=0 && a<=20 && b>=0 && b<=20 && c>=0 && c<=20 && dp[a][b][c] != 0){
            return dp[a][b][c];
        }
		
        if(a <= 0 || b <= 0 || c <= 0) {
            return 1;
        }
            
        if(a > 20 || b > 20 || c > 20) {
            return dp[20][20][20] = w(20, 20, 20);
        }
            
        if(a < b && b < c) {
            return dp[a][b][c] = w(a, b, c - 1) + w(a, b - 1, c - 1) - w(a, b - 1, c);
        }
        
        return dp[a][b][c] = w(a - 1, b, c) + w(a - 1, b - 1, c) + w(a - 1, b, c - 1) - w(a - 1, b - 1, c - 1);
    }

}

File: 백준/Silver/9184. 신나는 함수 실행/신나는 함수 실행.py
import sys
# sys.stdin = open('input.txt', 'r')
input = sys.stdin.readline

dp = [[[[] for _ in range(21)] for _ in range(21)] for _ in range(21)]


a,b,c = map(int,input().split())


for i in range(21):
    for j in range(21):
        dp[0][i][j] = 1
        dp[i][0][j] = 1
        dp[i][j][0] = 1
        
for i in range(1,21):
    for j in range(1,21):
        for k in range(1,21):
            if i<j and j<k:
                dp[i][j][k] = dp[i][j][k-1] + dp[i][j-1][k-1] - dp[i][j-1][k]
            else:
                dp[i][j][k] = dp[i-1][j][k] + dp[i-1][j-1][k] + dp[i-1][j][k-1] - dp[i-1][j-1][k-1]
def w(a,b,c):
    if a <= 0 or b <= 0 or c <= 0:
        return dp[0][0][0]
    if a > 20 or b > 20 or c > 20:
        return dp[20][20][20]
    else:
        return dp[a][b][c]
while a != -1 or b!= -1 or c != -1:
    print(f"w({a}, {b}, {c}) = {w(a,b,c)}")
    a,b,c = map(int,input().split())
    